docs/proposals/runtime-image.md

```diff
@@ -0,0 +1,122 @@
+
+# Motivation
+
+With the runtime-image feature in place, it will be employed for the creation of _lean-images_, which takes the ability of picking only certain parts of the original image, creating a runtime-image with the reduced data-set.
+
+Additionally, creating images with a different base-image, and other advanced use-cases are also supported by this feature.
+
+# Proposal
+
+The sections below are describing the changes necessary to introduce the runtime-image.
+
+## API Changes
+
+A new attribute will be added on Build resource, as the following example:
+
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+# ...
+spec:
+  runtime:
+    base:
+      image: runtime-base-image:latest
+    workDir: /workspace/source
+    env:
+      JAVA_HOME: /path/to/java-home
+    labels:
+      custom-image-label: value
+    directories:
+      - /path/to/directory
+      - /path/to/another/directory:/target/location
+    entrypoint:
+      - java
+      - -jar
+      - ...
+```
+
+The new attributes added are:
+
+- `spec.runtime.base`: specifies the runtime base-image to be used, using [Image](https://github.com/redhat-developer/build/blob/97012ab56417ce1691a70896d90e490ea6a4d23c/pkg/apis/build/v1alpha1/build_types.go#L58) as type;
+- `spec.runtime.workDir`: path to `WORKDIR` in runtime-image;
+- `spec.runtime.env`: runtime-image additional environment variables, key-value;
+- `spec.runtime.labels`: runtime-image additional labels, key-value;
+- `spec.runtime.directories`: list of directories paths to be copied to runtime-image, those can be defined as `<source>:<destination>` split by colon (`:`);
+- `spec.runtime.entrypoint`: entrypoint command, specified as a list;
+
+## Runtime-Image
+
+Given the API changes described above, we will now address how the runtime-image will be created. This image is always based on an image built by the `BuildStrategy` (or `ClusterBuildStrategy`) steps, and therefore, we can introduce new steps to add the elements needed for the runtime-image creation.
+
+
+### Runtime Dockerfile
+
+The first step to be appended on the given `BuildStrategy` instance is the `Dockerfile` for the runtime-image, commonly named `Dockerfile.runtime`.
+
+This Dockerfile would like the following example:
+
+
+```
+FROM output-image:tag as builder
+
+FROM runtime-image:tag
+ENV VARIABLE=value
+LABEL label=value
+COPY --from=builder "/path/to/source" "/path/to/destination"
+WORKDIR "/path/to/destination"
+ENTRYPOINT [ "command", "args" ]
+```
+
+The example shows a [multi-stage image build](https://docs.docker.com/develop/develop-images/multistage-build/) using the outcomes of the original `BuildStrategy` steps as `builder` image, and afterwards, rendering all the elements supported by the [new `runtime` section](#API-Changes). To create this file we may employ [`text/template`](https://golang.org/pkg/text/template/) to render its contents.
```

Comment:
Is my understanding correct that the intermediate image would be pushed to the destination, then used again in the runtime container build as `FROM` and then pushed again? If so, then we have two issues:

- There is a time window when the intermediate image is visible, if tag is something like `latest`, then this might be pulled.
- If the runtime container build fails (not necessarily due to a bug but potentially due to a timeout), the intermediate image would stay there.

I do not have a good solution proposal. One could use a special tag like `intermediate-${buildrunname}` but the end user would need to clean up this which is not nice.

Reply:
Good point, the ideal solution here would've been to not use the image from the registry - as @otaviof had initially suggested. 

But then, in that case, different strategies have different ways of storing the images in the file system, and that makes things a bit complicated?

Reply:
@sbose78 yeah, I know. That's also what I thought about when you said you wanted to take the `push` out of the strategy itself. For reference, that issue is here: [Remove image push steps from the build strategies #165](https://github.com/redhat-developer/build/issues/165).

Reply:
Decision has been made in today's community meeting to stick to the current behavior = two images will be pushed to the destination. First, the image output from the build strategy, second, the runtime image by the generated kaniko step. Documentation should be added to explain this behavior. At a later point, we will want to address this when we try to remove the image push from the build strategy.